pkg/controller/uniteddeployment/adapter/statefulset_adapter.go

```diff
@@ -0,0 +1,260 @@
+package adapter
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	alpha1 "github.com/openkruise/kruise/pkg/apis/apps/v1alpha1"
+	kruisectlutil "github.com/openkruise/kruise/pkg/controller/util"
+	"github.com/openkruise/kruise/pkg/util/refmanager"
+)
+
+type StatefulSetAdapter struct {
+	client.Client
+
+	Scheme *runtime.Scheme
+}
+
+// NewResourceObject creates a empty StatefulSet object.
+func (a *StatefulSetAdapter) NewResourceObject() runtime.Object {
+	return &appsv1.StatefulSet{}
+}
+
+// NewResourceListObject creates a empty StatefulSetList object.
+func (a *StatefulSetAdapter) NewResourceListObject() runtime.Object {
+	return &appsv1.StatefulSetList{}
+}
+
+// GetObjectMeta returns the ObjectMeta of the subset of StatefulSet.
+func (a *StatefulSetAdapter) GetObjectMeta(obj metav1.Object) *metav1.ObjectMeta {
+	return &obj.(*appsv1.StatefulSet).ObjectMeta
+}
+
+// GetStatusObservedGeneration returns the observed generation of the subset.
+func (a *StatefulSetAdapter) GetStatusObservedGeneration(obj metav1.Object) int64 {
+	return obj.(*appsv1.StatefulSet).Status.ObservedGeneration
+}
+
+// GetReplicaDetails returns the replicas detail the subset needs.
+func (a *StatefulSetAdapter) GetReplicaDetails(obj metav1.Object, updatedRevision string) (specReplicas, specPartition *int32, statusReplicas, statusReadyReplicas, statusUpdatedReplicas, statusUpdatedReadyReplicas int32, err error) {
+	set := obj.(*appsv1.StatefulSet)
+	var pods []*corev1.Pod
+	pods, err = a.getStatefulSetPods(set)
+	if err != nil {
+		return
+	}
+
+	specReplicas = set.Spec.Replicas
+	if set.Spec.UpdateStrategy.Type == appsv1.OnDeleteStatefulSetStrategyType {
+		revision := getRevision(&set.ObjectMeta)
+		specPartition = getCurrentPartition(pods, revision)
+	} else if set.Spec.UpdateStrategy.RollingUpdate != nil &&
+		set.Spec.UpdateStrategy.RollingUpdate.Partition != nil {
+		specPartition = set.Spec.UpdateStrategy.RollingUpdate.Partition
+	}
+
+	statusReplicas = set.Status.Replicas
+	statusReadyReplicas = set.Status.ReadyReplicas
+	statusUpdatedReplicas, statusUpdatedReadyReplicas = calculateUpdatedReplicas(pods, updatedRevision)
+
+	return
+}
+
+// GetSubsetFailure returns the failure information of the subset.
+// StatefulSet has no condition.
+func (a *StatefulSetAdapter) GetSubsetFailure() *string {
+	return nil
+}
+
+// ConvertToResourceList converts StatefulSetList object to StatefulSet array.
+func (a *StatefulSetAdapter) ConvertToResourceList(obj runtime.Object) []metav1.Object {
```

Comment:
I doubt if we need to expose this as interface method. Can we move this to subset_control.go as a private method?

Reply:
Use reflect, instead.
Please take a look in `subset_control.go` L59.